### DISCUSSION TITLE
How to handle transitive polyglot dependencies?

### DISCUSSION THREAD

**Discussion by rdicroce** [DependencyTrack/dependency-track#2648] [Q&A]
I have the following scenario:

- Maven (Java) project A depends on Maven project B.
- Project B embeds project C, an Angular (Webpack) frontend, into its JAR.
- Project C includes numerous NPM packages.

What I want is for Dependency-Track to treat vulnerabilities in any of those NPM packages as a vulnerability in project A. Is there any way to do that right now? After spending a lot of time yesterday searching and thinking about it, I couldn't find any.

It seems to me that there's a larger problem here, which is that there is currently no guidance from the DT docs or the CycloneDX spec about how to handle transitive dependencies. The spec is very flexible, but that creates a problem for DT, as seen in https://github.com/DependencyTrack/dependency-track/issues/869. I started this as a separate discussion because I don't want to derail that issue, and this isn't quite the same thing.

Currently, the official tools (cyclonedx-maven-plugin, cyclonedx-webpack-plugin, etc) produce SBOMs that include all transitive dependencies, but only from within the same ecosystem. I agree that including all transitive deps in a SBOM seems like the right solution, since many build systems let you exclude or change versions of transitive deps. But this raises the question of how to cross polyglot boundaries.

One possibility would be to stop including transitive deps in the SBOM, and instead produce separate SBOMs that are then linked together. IMO this is not the right answer, because it runs into that problem of excluded dependencies I mentioned above. Also, the CycloneDX spec provides for externalReferences to other SBOMs, but this means that either all of the SBOMs have to be uploaded into DT (problematic for third-party polyglot situations, such as sqlite-jdbc) or DT has to fetch the linked SBOMs (and I'm sure you can already see problems with restricted networks, dead URLs, etc).

Getting the dependencies used in project C into the SBOM for project A is borderline impossible right now. c-m-p will add project B to the SBOM, but project C is not a Maven project, so c-m-p is completely oblivious to it. I think there is a solution to this, but it's not simple.

Step one: enhance c-m-p so it can merge SBOMs together, like what cyclonedx-cli can do. That way, project B can merge project C's SBOM into its own SBOM during the build process.

Step two: some sort of marker is added to project B's SBOM to distinguish "fixed" versus "floating" components. (Alternative terminology: "embedded" versus "resolved".) Project B's dependencies on all those NPM packages is fixed, because those are embedded into the JAR. They're not Maven dependencies, so there is no way for project A to exclude them or pick a different version. But if project B had a Maven dependency on something else, e.g. a logging API, that dependency could be excluded or overridden, so that is a floating dependency.

Step three: enhance c-m-p so that when it resolves dependencies, it also checks to see if the Maven repo contains a SBOM for that dependency. If yes, it fetches the dependency's SBOM and looks for fixed dependencies. Those are then added into project A's SBOM.

I've used Maven and NPM in this example, but you can see how this could be generalized to work with other ecosystems. None of this requires a change to the CycloneDX spec, except maybe for the part about the fixed/floating marker (but that could also be done using the properties element). The resulting SBOM for project A would be complete and correct, and would remove the need for DT to figure out how to link these projects together. I'd be willing to work on the changes to c-m-p if we can get consensus on this.

Thoughts?

**Comment by fmarot** [comment]
Hello,
just so you know you are not alone in this kind of situation : I also have a java/Maven/multi-modules project which embeds a javascript app. So I just produce 2 SBOMS, one for each ecosystem, generated by their respective build tools (Maven & npm) and then I automatically merge both using cyclonedx-cli (and then, I upload it to a dependency-track server fwiw). It does the job for the moment.

For the record I also had an interesting discussion about this on the slack channel, it may help: https://cyclonedx.slack.com/archives/CVCKP34A2/p1663172830260619
